README
Welcome to the WebGL Action RPG GitHub's page !

This is a no-class action role playing game (RPG). See game mechanism in the wiki.

The online demo is available here : http://nicolas-bonnel.github.com/WARPG/index.html

This game requires a webGL compatible web browser. It has only been tested with chrome 10 and firefox 4. It may work with older versions of chrome, and some features do not work with firefox 4.0 (particle system based on sprites). You will probably need to download a beta of a recent browser to see this demo, see http://khronos.org/webgl/wiki/Getting_a_WebGL_Implementation"

If you need information about this demo, you can contact me : nicolas dot bonnel at gmail dot com. The code is licenced under GNU GPL 3. Feel free to reuse the code, credits would be appreciated.

Any help would be appreciated, see the wiki to see how you to contribute.

CHANGELOG
v 0.06
- New items : leather armor parts
- Modified the inventory : new picture for the hero and stats displayed (damages and resists)
- Added resists on armors
- Skills now use equipment damage modifiers
- Removed some skills : magic missile and melee attack
- New skill : ice nova, fire missiles, charged bolts, slashing, piercing and blunt attacks.
- Reworked other skills a little bit
- Added antlion bosses in forests
- Added a trader that can buy your items.
- Added a makefile in the project : all js files are now concatened into one file. This makes the gh-pages branch cleaner.

v 0.05 (04/04/2011)
- New items : weapons, armors parts
- Items have a quality : normal, magic (1-2 powers), rare (3-4 powers) and epic (5-6 powers)
- Few powers availables for items : +elemental damage, + HP/SP/MP
- Added chests (A to open) that drop higher quality loot
- Added a Trash in the inventory menu
- Added / modfied new 3D objects : trees and training stuff
- Modified a little bit some monster stats

v 0.04 (03/22/2011)
- Added a new monster : spider (thanks to Wciow and John.d.h).
- New model for the hero : a goth woman (thanks to TiZiana). The quality of animations is quite poor: this is my first attempt at animating a 3D model :P. I'll soon try to improve her animations, and i'll upload the result on OGA.
- Started to work on the inventory. You can now switch weapon with the ones you dropped and drag potions on the shortcut bar.
- Monsters can now drop items. Added new items (katana, staff, mace, HP, SP and MP potions). Potions stack and give 10 units back.
- Reduced the map size: the previous maps were a little too big, causing overflows in vertices buffers.
- Improved transparency: trees should have a better look.
- Started to use jquery : skills and items should now be draggable with all recent browsers.
- The description of skills and items is now shown when the mouse is over the icon.
- Improved the pause a little bit.
- Now a 'miss' appear when dodging attacks.

v 0.03 (03/05/2011)
- Removed the map file, now the map is randomly generated with square & diamond algorithm and is a little bit larger.
- Switched matrix library to glUtils (link is in the credit page) which is a lot faster.
- Added an area concept : the map is composed of various areas linked together by paths. Villages need some improvement thought.
- Added a starting point, with a shrine, when you die you will revive there. Monsters level is now given by the level of an area, which increase with the distance in the graph of the starting point.
- Added cemetery objects, this still need some improvements to make nice graveyards. Added skeleton enemies.
- The drawback of these updates : loading time has increase (between 5 and 10 sec). I'll add a loading screen soon.

v 0.02 (02/27/2011)
- Removed the 'Q' shortcut for fireball.
- Skills have now a level requirement.
- Added a skill panel : you can access it by clicking on an aptitude in the aptitude panel ('L' shortcut). It then will propose all skills linked to this aptitude that you can use according to your aptitude levels. You can then drag & drop skills on the shortcut bar.
- Added a dozen of skills : melee (precise attack, power attack, burning strike, thunder strike), fire (flare, inferno, flame nova), air (lightning bolt, lightning, thunder, nova).

v 0.01 (02/15/2011)
- You can now die and monsters disapear when dead. Added some randomness in monster level (increases with the distance to (0,0))
- Added 2 panels (shortcuts 'P' and 'L') : settings (only view radius for the moment) and aptitudes.
- You can now level up some aptitudes (constitution, intelligence, stamina, dexterity, melee, cast fire, sword).
- When hit you enter in recovery mode, which cancel your current action.
- Added a particule system. Particle effects are written in json files which should allow quick particle definition
- Added the fireball skill ('Q' shortcut) :)

v 0.00 (01/25/2011)
- Added sky, fog and a grid system for optimization, there is now 2000 objects on the map and the demo should run on lower hardware.
- Added static models (chest, house, shrine, trees). Added monsters : minotaur and antlion (special thanks to pfunked : you made awesome stuff).
- Linked a sword to the skeleton model : you can now attack monsters.
- Added collision detection
- Monsters have an health disk bellow them and can now die.
- Improved drawing (objects in the view radius but in the back are not drawn anymore)


js/item.js
/* Javascript Item class
 *
 * Copyright (C) 2010   Nicolas Bonnel (nicolas.bonnel@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

var numPower = {};
numPower['normal'] = "0";
numPower['magic'] = "1+Math.floor(rand()*2)";
numPower['rare'] = "3+Math.floor(rand()*2)";
numPower['epic'] = "5+Math.floor(rand()*2)";

var priceModif = {};
priceModif['normal'] = 1;
priceModif['magic'] = 2;
priceModif['rare'] = 4;
priceModif['epic'] = 8;

function Item(jsonItem,quality,itemLevel){
	if(jsonItem.modelName){
		//debug(jsonItem.modelName);
		var item = new Object3D(0,0,0,jsonItem.modelName);
	}else
		var item = new Object3D(0,0,0,null);
	item.name = jsonItem.name;
	item.type = jsonItem.type;
	item.requirements = jsonItem.requirements;
	if(jsonItem.resists)
		item.resists = jsonItem.resists;
	if(item.type=='weapon'){
		item.damageType = jsonItem.damageType;
		item.damages = jsonItem.damages;
		item.damageRange = jsonItem.damageRange;
		item.range = jsonItem.range;
	}
	if(jsonItem.goldCost){
		item.goldCost = jsonItem.goldCost*priceModif[quality];
		item.description = jsonItem.description;
		item.icon = jsonItem.icon;
	}else
		item.goldCost = 0;
	if(jsonItem.stackable)
		item.stackable = jsonItem.stackable;
	item.id = rand()+item.name+rand();
	if(item.type == 'consumable'){
		item.quality = 'normal';
	}else{
		item.quality = quality;
	}
	item.level = jsonItem.level;
	//debug(item.name+' : '+item.level);
	item.powers = {};
	for (var i=0;i<eval(numPower[item.quality]);i++){
		var tmpPow = powers.list[Math.floor(rand()*powers.list.length)];
		if(item.powers[tmpPow.name]){
			//debug(tmpPow.name+' : '+item.powers[tmpPow.name].effect.value);
			item.powers[tmpPow.name].effect.value = Math.round(10*(eval(tmpPow.effect['value'])+item.powers[tmpPow.name].effect.value))/10;
		}else{
			var power = {};
			power.effect = {};
			for (var x in tmpPow.effect){
				if (x=='value')
					power.effect[x] = Math.round(10*eval(tmpPow.effect[x]))/10;
				else
					power.effect[x] = tmpPow.effect[x];
			}
			item.powers[tmpPow.name] = power;
		}
	}
	return item;
}

function getIcon(item){
	var ret = $('<div></div>').attr({'class':'item','id':item.id,'name':item.name,'style':'width:32;height:32;background-image:url(data/icons/'+item.icon+')'}).draggable({revert:'invalid',appendTo: 'body',helper: 'clone',start:function(){src = $(this).parent();}});
	ret.mouseover(function() {
	    $('#canvas3D').append(itemDescription(world.player.inventory[$(this).attr('id')]));
	  }).mouseout(function(){
	    $('.description').remove();
	  });
	if(item.stackable){
		ret.append('1');
		ret.attr({'count':1});
	}
	ret.addClass(item.type);
	if(items[item.name].type && items[item.name].type=='consumable'){
		ret.click(function(){
			var effect = items[$(this).attr('name')].effect;
			if(effect.recover)
				world.player.recover(effect.recover[0],effect.recover[1],effect.recover[2]);
			$(this).attr('count',eval($(this).attr('count'))-1);
			$(this).empty();
			$(this).append($(this).attr('count'));
			if($(this).attr('count')=='0'){
				$(this).remove();
				delete world.player.inventory[$(this).attr('id')];
			}
		});
	}
	return ret;
}

function itemDescription(item){
	setAptitudeVars(world.player);
	var descrip ='<div class="description" style=left:'+mouseX+';top:'+mouseY+';><font color="#';
	if(item.quality=='normal')
		descrip += 'FFFFFF';
	else if(item.quality=='magic')
		descrip += 'AAAAFF';
	else if(item.quality=='rare')
		descrip += 'AAFFAA';
	else if(item.quality=='epic')
		descrip += 'FFDDAA';
	descrip +='">'+ item.name+' ('+item.type+')</font><br>'+item.description+'<br>';
	if(item.type=='weapon')
		descrip += 'Damages ('+item.damageType+') : '+(item.damages-item.damages*item.damageRange)+'-'+(item.damages+item.damages*item.damageRange)+', range : '+item.range+'<br>';
	for (var x in item.powers){
		descrip += '+ '+item.powers[x].effect.value+' ';
		if (item.powers[x].effect.type == 'damage')
			descrip += item.powers[x].effect.damageType+ ' damage<br>';
		else if (item.powers[x].effect.type == 'bonus')
			descrip += item.powers[x].effect.aptitude+'<br>';
	}
	if(item.resists){
		descrip += 'Resists : ';
		for (var x in item.resists)
			descrip += x+' : '+item.resists[x]+' %, ';
		descrip = descrip.substring(0,descrip.length-2)+'<br>';
	}
	if (item.goldCost>0)
		descrip += '<font color="#EEEEAA">Price: '+item.goldCost+' gold.</font>';
	descrip += '</div>';
	return descrip;
}

function loot(level,qual){
	var itemNames = [];
	for (x in items)
		if (items[x].droppable && (!items[x].level || items[x].level <= level))
			itemNames.push(items[x].name);
	var quality;
	if (qual<0.01)
		quality = 'epic';
	else if (qual<0.05)
		quality = 'rare';
	else if (qual<0.2)
		quality = 'magic';
	else
		quality = 'normal';
	var it = new Item(items[itemNames[Math.floor(rand()*itemNames.length)]],quality,level);
	it.pickable = new ParticleEmitter(it,'itemdrop');
	world.particles.push(it.pickable);
	return it;
}
